### ISSUE TITLE
SVD gradient is unstable for non-unique singular values

### ISSUE THREAD

**Issue by christian-rauch** [tensorflow/tensorflow#17476]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 14.04
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version (use command below)**: ('v1.5.0-10-g5b10b34', '1.5.0')
- **Python version**: Python 2.7.6
- **Bazel version (if compiling from source)**: 0.11.0
- **GCC/Compiler version (if compiling from source)**: 4.8.5
- **CUDA/cuDNN version**: 8.0 / 6.0
- **GPU model and memory**: GeForce GTX 970 4GB
- **Exact command to reproduce**:

### Describe the problem
The gradient of SVD becomes `nan` when singular values are not unique.

From the code
https://github.com/tensorflow/tensorflow/blob/c6a12c77a50778e28de3590f4618bc2b62f3ecab/tensorflow/python/ops/linalg_grad.py#L332-L342
it is clearly visible that `f` becomes `inf` for equal singular values.

I briefly skimmed through the paper https://arxiv.org/abs/1509.07838 but couldn't find an explanation for this behaviour. Can someone give an intuitive example why the gradient for similar singular values should not be defined?

Similar singular values commonly appear for estimation of rotation matrices (all singular values become 1). At the moment it is impossible to use SVD for such a case.

Alternative implementations of the SVD gradients based on the same paper are:
- https://github.com/InhaDeeplearningGroup/Academic_research/blob/master/LSH/tensorflow_slim/svdGradients.py
- https://gist.github.com/psycharo/60f58d5435281bdea8b9d4ee4f6e895b

which circumvent this problem by simply replacing the `nan`s or effectively setting them to 0.

Could this be used alternatively to the current implementation that just divides by 0? Or alternatively just add a small epsilon to the difference of singular values.

Btw, I think that no gradient computation in TensorFlow should ever return faulty gradients but raise an error message to simplify debugging. I don't see the point of using `nan` or `inf` gradients.

**Comment by yaroslavvb** [comment]
Consider an identity matrix where SVD reduces to eigendecomposition. Every vector is an eigenvector. You have to pick a single output vector in order to be able to backprop through it, but because all choices are equally valid, the backprop is not defined.

**Comment by shoyer** [comment]
> which circumvent this problem by simply replacing the nans or effectively setting them to 0.

This at best patching over the problem. I don't think it would actually solve the problem for training deep learning models. You will still have very large gradients that lead to numerical stability problems when there are nearly degenerate singular values.

**Comment by shoyer** [comment]
> Similar singular values commonly appear for estimation of rotation matrices (all singular values become 1). At the moment it is impossible to use SVD for such a case.

Rotation matrices also suffer from gimbal lock (which I think may be related to this issue of degenerate singular values). You probably want to use [quaternions for spatial rotation](https://en.wikipedia.org/wiki/Quaternions_and_spatial_rotation) instead.

**Comment by christian-rauch** [comment]
Thanks for you explanation.
If a gradient is not defined, wouldn't it be better to provide no parameter updates at all (i.e. return a gradient of 0)? At the moment, the gradient becomes undefined at some circumstances (non-unique singular values) and propagates further, which breaks the whole training process without further warnings.

**Comment by yaroslavvb** [comment]
It's a similar to having gradients of 1/x or sqrt(x). If gets close to 0, it breaks your training. Traditionally it's been up to the user to make sure that doesn't happen.

**Comment by christian-rauch** [comment]
I am mostly concerned with the application of rotation estimation via SVD. After reading through https://arxiv.org/pdf/0904.1613.pdf I recognised that the underlying problem is the uniqueness of the SVD solution for the optimal R. I think the gimbal lock problem is related to Euler angles, not rotation matrices. E.g. rotation estimation via quaternions involves choosing the eigenvector with the largest eigenvalue which also becomes non-unique when all eigenvalues are equal.
